File: phases/19-capstone-projects/09-build-a-container-runtime/code/main.c
// Build a Container Runtime — Namespace Setup (C)
// Run: gcc -o nsenter main.c && ./nsenter <rootfs> <command> [args...]
// Note: Linux only — requires clone(), _GNU_SOURCE, sched.h
//
// Uses clone() with CLONE_NEWPID | CLONE_NEWNS | CLONE_NEWUTS to create
// isolated namespaces, mount procfs, set hostname, and exec the container command.

#define _GNU_SOURCE
#include <sched.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/mount.h>
#include <sys/wait.h>
#include <signal.h>
#include <errno.h>
#include <string.h>

// Set up a minimal container environment
int container_child(void *arg) {
    char **argv = (char **)arg;

    // Set hostname
    sethostname("container", 9);

    // Mount proc filesystem (requires PID namespace)
    if (mount("proc", "/proc", "proc", 0, NULL) == -1) {
        fprintf(stderr, "mount proc: %s\n", strerror(errno));
    }

    // Set environment
    setenv("container", "true", 1);

    // Execute the container command
    execvp(argv[0], argv);
    perror("execvp failed");
    return 1;
}

#define STACK_SIZE (1024 * 1024) // 1MB child stack

int main(int argc, char *argv[]) {
    if (argc < 3) {
        fprintf(stderr, "Usage: %s <rootfs> <command> [args...]\n", argv[0]);
        fprintf(stderr, "\nExample: %s / /bin/sh -c 'echo hello from container'\n", argv[0]);
        return 1;
    }

    char *rootfs = argv[1];
    char **cmd = &argv[2];

    // Allocate stack for child
    char *stack = malloc(STACK_SIZE);
    if (!stack) { perror("malloc"); return 1; }

    // Clone with new PID and mount namespaces
    int flags = CLONE_NEWPID | CLONE_NEWNS | CLONE_NEWUTS | SIGCHLD;

    pid_t child = clone(container_child, stack + STACK_SIZE, flags, cmd);
    if (child == -1) {
        perror("clone");
        free(stack);
        return 1;
    }

    printf("Container started with PID %d (host), PID 1 (container)\n", child);

    // Wait for container to exit
    int status;
    waitpid(child, &status, 0);

    if (WIFEXITED(status)) {
        printf("Container exited with status %d\n", WEXITSTATUS(status));
    } else if (WIFSIGNALED(status)) {
        printf("Container killed by signal %d\n", WTERMSIG(status));
    }

    free(stack);
    return 0;
}

File: phases/19-capstone-projects/09-build-a-container-runtime/code/main.go
// Build a Container Runtime
// Run: go build -o minicontainer main.go && ./minicontainer spec.json
// Note: Linux only — uses clone(), cgroups v2, namespace flags
//
// Architecture:
//   Container Spec (JSON) → Cgroup Setup → Namespace Clone → Process Start → Wait
//
// Implements a minimal container runtime with namespace isolation and cgroup limits.

package main

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"syscall"
)

// =============================================================================
// Step 1: Container Spec Parser
// =============================================================================

type ContainerSpec struct {
	Rootfs   string            `json:"rootfs"`
	Args     []string          `json:"args"`
	Env      map[string]string `json:"env"`
	Hostname string            `json:"hostname"`
	Limits   ResourceLimits    `json:"limits"`
}

type ResourceLimits struct {
	MemoryLimitMB int `json:"memory_limit_mb"`
	CPUShares     int `json:"cpu_shares"`
	PidLimit      int `json:"pid_limit"`
}

func loadSpec(path string) (*ContainerSpec, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("reading spec: %w", err)
	}
	var spec ContainerSpec
	if err := json.Unmarshal(data, &spec); err != nil {
		return nil, fmt.Errorf("parsing spec: %w", err)
	}
	return &spec, nil
}

// =============================================================================
// Step 2: Cgroup Setup
// =============================================================================

const cgroupRoot = "/sys/fs/cgroup"

func setupCgroup(containerID string, limits ResourceLimits) (string, error) {
	cgroupPath := filepath.Join(cgroupRoot, "mini-container", containerID)

	if err := os.MkdirAll(cgroupPath, 0755); err != nil {
		return "", fmt.Errorf("creating cgroup: %w", err)
	}

	if limits.MemoryLimitMB > 0 {
		limitBytes := limits.MemoryLimitMB * 1024 * 1024
		memPath := filepath.Join(cgroupPath, "memory.max")
		if err := os.WriteFile(memPath, []byte(strconv.Itoa(limitBytes)), 0644); err != nil {
			return "", fmt.Errorf("setting memory limit: %w", err)
		}
	}

	if limits.CPUShares > 0 {
		cpuPath := filepath.Join(cgroupPath, "cpu.weight")
		if err := os.WriteFile(cpuPath, []byte(strconv.Itoa(limits.CPUShares)), 0644); err != nil {
			return "", fmt.Errorf("setting CPU shares: %w", err)
		}
	}

	if limits.PidLimit > 0 {
		pidPath := filepath.Join(cgroupPath, "pids.max")
		if err := os.WriteFile(pidPath, []byte(strconv.Itoa(limits.PidLimit)), 0644); err != nil {
			return "", fmt.Errorf("setting PID limit: %w", err)
		}
	}

	return cgroupPath, nil
}

func addToCgroup(cgroupPath string, pid int) error {
	procsPath := filepath.Join(cgroupPath, "cgroup.procs")
	return os.WriteFile(procsPath, []byte(strconv.Itoa(pid)), 0644)
}

func cleanupCgroup(cgroupPath string) error {
	return os.RemoveAll(cgroupPath)
}

// =============================================================================
// Step 3: Container Runtime
// =============================================================================

func runContainer(spec *ContainerSpec) error {
	containerID := "container-001"

	cgroupPath, err := setupCgroup(containerID, spec.Limits)
	if err != nil {
		return fmt.Errorf("cgroup setup: %w", err)
	}
	defer cleanupCgroup(cgroupPath)

	cmd := exec.Command(spec.Args[0], spec.Args[1:]...)
	cmd.Stdin = os.Stdin
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	cmd.SysProcAttr = &syscall.SysProcAttr{
		Cloneflags: syscall.CLONE_NEWPID | syscall.CLONE_NEWNS | syscall.CLONE_NEWUTS,
	}

	env := []string{"container=true", fmt.Sprintf("hostname=%s", spec.Hostname)}
	for k, v := range spec.Env {
		env = append(env, fmt.Sprintf("%s=%s", k, v))
	}
	cmd.Env = env

	if err := cmd.Start(); err != nil {
		return fmt.Errorf("starting container: %w", err)
	}

	fmt.Printf("Container %s started (PID %d)\n", containerID, cmd.Process.Pid)

	if err := addToCgroup(cgroupPath, cmd.Process.Pid); err != nil {
		cmd.Process.Kill()
		return fmt.Errorf("adding to cgroup: %w", err)
	}

	return cmd.Wait()
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: minicontainer <spec.json>")
		fmt.Println("\nExample spec.json:")
		fmt.Println(`  {"rootfs":"/","args":["/bin/sh","-c","echo hello"],"env":{},"hostname":"mycontainer","limits":{"memory_limit_mb":128,"cpu_shares":1024,"pid_limit":100}}`)
		os.Exit(1)
	}

	spec, err := loadSpec(os.Args[1])
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error loading spec: %v\n", err)
		os.Exit(1)
	}

	if err := runContainer(spec); err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}
}
